test/test_dispatch.py

```diff
@@ -0,0 +1,359 @@
+        ], raises=True)
+        self.assertExpectedInline(r, '''In registration for test::foo: expected schema of operator to be "test::foo(Tensor x, Tensor y) -> (Tensor)", but got inferred schema "(Tensor _0) -> (Tensor _0)". The number of arguments is different. 2 vs 1.''')  # noqa
+
+    def test_def_with_inference(self):
+        r = self.commute("foo", [
+            # m.def("foo", [](const Tensor & x) { return x })
+            lambda m: m.def_name_t_t("foo"),
+            # m.impl("foo", torch::dispatch_autograd([](const Tensor & x) { return x }))
+            lambda m: m.impl_t_t("foo", "autograd")
+        ])
+        self.assertExpectedInline(r, '''\
+name: test::foo
+schema: test::foo(Tensor _0) -> (Tensor _0)
+alias analysis kind: CONSERVATIVE
+VariableTensorId: impl_t_t :: (Tensor _0) -> (Tensor _0) [ boxed unboxed ]
+catchall: default_def_name_t_t :: (Tensor _0) -> (Tensor _0) [ boxed unboxed ]
+''')
+
+    def test_def_only(self):
+        r = self.commute("foo", [
+            # m.def("foo(Tensor x, Tensor y) -> Tensor")
+            lambda m: m.def_("foo(Tensor x, Tensor y) -> Tensor"),
+        ])
+        self.assertExpectedInline(r, '''\
+name: test::foo
+schema: test::foo(Tensor x, Tensor y) -> (Tensor)
+alias analysis kind: FROM_SCHEMA
+''')
+
+    def test_impl_only(self):
+        r = self.commute("foo", [
+            # m.impl("foo", [](const Tensor& x) { return x })
+            lambda m: m.impl_t_t("foo"),
+            # m.impl("foo",
+            #        torch::dispatch_autograd([](const Tensor& x) { return x }))
+            lambda m: m.impl_t_t("foo", "autograd")
+        ])
+        self.assertExpectedInline(r, '''\
+name: test::foo
+schema: (none)
+VariableTensorId: impl_t_t :: (Tensor _0) -> (Tensor _0) [ boxed unboxed ]
+catchall: impl_t_t :: (Tensor _0) -> (Tensor _0) [ boxed unboxed ]
+''')
+
+    # Can't do this yet for BC reasons
+    @unittest.expectedFailure
+    def test_multiple_def_error(self):
+        r = self.commute("foo", [
+            # m.def("foo(Tensor x, Tensor y) -> Tensor")
+            lambda m: m.def_("foo(Tensor x, Tensor y) -> Tensor"),
+            # m.def("foo(Tensor x, Tensor y) -> Tensor")
+            lambda m: m.def_("foo(Tensor x, Tensor y) -> Tensor"),
+        ], raises=True)
+        # TODO: fill in the error message here
+        # self.assertExpectedInline(r, '''''')
+
+    def test_def_with_explicit_alias(self):
+        r = self.commute("foo", [
+            # m.def(torch::schema(
+            #   "foo(Tensor x, Tensor y) -> Tensor",
+            #   AliasAnalysisKind::PURE))
+            lambda m: m.def_("foo(Tensor x, Tensor y) -> Tensor",
+                             alias="PURE_FUNCTION")
+        ])
+        self.assertExpectedInline(r, '''\
+name: test::foo
+schema: test::foo(Tensor x, Tensor y) -> (Tensor)
+alias analysis kind: PURE_FUNCTION
+''')
+
+    # TODO: get rid of this test when multiple defs are wrong
+    def test_multiple_def_schema_mismatch(self):
```

Comment:
You can probably add a few tests for overload names, i.e. multiple defs are valid if different overload names.

Reply:
I'm going to add this as a C++ test

Reply:
https://github.com/pytorch/pytorch/pull/35061/commits/ab6ab1ebc298f4a40dc808e61025d8f40c15c3d0